From patchwork Fri Aug 10 21:26:02 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Drew Schmitt <dasch@google.com>
X-Patchwork-Id: 10563147
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7E5C2157B
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:12 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6F2372B5B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:12 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 6274F2B5BF; Fri, 10 Aug 2018 21:26:12 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 0BE3A2B5B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:12 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727239AbeHJX5p (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 10 Aug 2018 19:57:45 -0400
Received: from mail-qt0-f201.google.com ([209.85.216.201]:32780 "EHLO
        mail-qt0-f201.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727206AbeHJX5p (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 10 Aug 2018 19:57:45 -0400
Received: by mail-qt0-f201.google.com with SMTP id j11-v6so8459957qtp.0
        for <kvm@vger.kernel.org>; Fri, 10 Aug 2018 14:26:09 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=1ec3eBMJbKj28anjPsfDGa8Y0bg9STFxgK929iMKzaI=;
        b=K/mxDeCbeOAATLqPzS9iIxjCr+ApEUURYpFMV84XkA1mvHoKUxYjYbOdjpXSKHiBjy
         10dKPcLsjrkuySsQS1zkyYsmYIomD5cOQ0g+U3clY2X1/qAzEaTitvn84vFqfnNutAqa
         ncFPCxbB7hlhpqEAer4uScQyaNaazu8BbiC1XV4yOaotm/o7OAep3twlqupifZSs9Ptw
         Uq4ETG3SpAWZ3+AsUrHxG87C+9S+CDjXXWQPKrREJK8hZfVLs/n0m1gk8aiMPtnzd6YK
         dvbkB2dILusE+F0y2Ej7pEzrWdPE5UUwURlDQZ2mbzvKtjoiqs8MuwKmx879IWZ4TY1n
         1Cyg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=1ec3eBMJbKj28anjPsfDGa8Y0bg9STFxgK929iMKzaI=;
        b=YQ/y0n+L1mWMb3M07eihZFmA7Q5q+3Csr+H3UeTWh4zMEDm95h7yOcc/GPlBuGWrfe
         NYAEpfTj4ORsa181Kee0QcWrJKEEM46/SbE68aHkV1ZNIZDgCLjRAjcK2zG5Kk/L8fkk
         5Bhd2GZJrw8+R77gG2I1WD7kbzT2FmOR7Op3ERLToOy7z88secZ0QqjfpFifh/Dd9Ywh
         6pDuevTDdkAfs6cgiR6x6fQPCVi+SRUgZnAq9l+bfKUC0spj3T2bW5Fleza4KvuWM1Rm
         GpMCpfAo1LP+nEq+XnCDqrZ0Rt5PZpxHzvab6FQ5a9cEXo5imHoCtQ4dfrCEevCAp7vr
         WZtw==
X-Gm-Message-State: AOUpUlHhtk9nlItly8LtR5jQE0bJ6iEYce0bmrt63mve/D/JNzkmzsDl
        5NQyOMAcQbBhjZFT5w3UfnWYrdXvQ994eS1lrZXi3fScVYtMRkNgflxD8YP3yrNyYmdQoyFoZ+R
        Zf7wBzVmAwWaCs0XiQcU1xfXfqeh+BheC4xlfLuX06ROQeK84dOOf4sY=
X-Google-Smtp-Source: 
 AA+uWPzlr8vTqE1Qys/RtQPFBseG/Rn66cKyeQrXcWmHVVqo8TQzgnk586+Fo2QgUetSbwbbvT6pNOOieA==
X-Received: by 2002:aed:3bd1:: with SMTP id
 s17-v6mr4437730qte.24.1533936369192;
 Fri, 10 Aug 2018 14:26:09 -0700 (PDT)
Date: Fri, 10 Aug 2018 14:26:02 -0700
Message-Id: <20180810212604.148650-1-dasch@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.18.0.597.ga71716f1ad-goog
Subject: [PATCH 1/3] KVM: x86: Turbo bits in MSR_PLATFORM_INFO
From: Drew Schmitt <dasch@google.com>
To: kvm@vger.kernel.org
Cc: Drew Schmitt <dasch@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Allow userspace to set turbo bits in MSR_PLATFORM_INFO. Previously, only
the CPUID faulting bit was settable. But now any bit in
MSR_PLATFORM_INFO would be settable.

Signed-off-by: Drew Schmitt <dasch@google.com>
---
 arch/x86/kvm/x86.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 0046aa70205aa..0278cde1893ca 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -2503,7 +2503,6 @@ int kvm_set_msr_common(struct kvm_vcpu *vcpu, struct msr_data *msr_info)
 		break;
 	case MSR_PLATFORM_INFO:
 		if (!msr_info->host_initiated ||
-		    data & ~MSR_PLATFORM_INFO_CPUID_FAULT ||
 		    (!(data & MSR_PLATFORM_INFO_CPUID_FAULT) &&
 		     cpuid_fault_enabled(vcpu)))
 			return 1;

From patchwork Fri Aug 10 21:26:03 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Drew Schmitt <dasch@google.com>
X-Patchwork-Id: 10563149
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 1D21C157B
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:16 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 0F62E2B5B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:16 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 024F62B5BF; Fri, 10 Aug 2018 21:26:15 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 99F992B5B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:15 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727280AbeHJX5t (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 10 Aug 2018 19:57:49 -0400
Received: from mail-oi0-f73.google.com ([209.85.218.73]:42953 "EHLO
        mail-oi0-f73.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727123AbeHJX5t (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 10 Aug 2018 19:57:49 -0400
Received: by mail-oi0-f73.google.com with SMTP id l14-v6so10528737oii.9
        for <kvm@vger.kernel.org>; Fri, 10 Aug 2018 14:26:13 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=Qpoq8njNIKPGqHkO3kbGHXyyjepVG6Du3/cu/4cpQMo=;
        b=CP4JCFPNvJinUmaU64tQ3ulsKqFh7jzgJyJXFXnkewHsV3PCUg+P3LvuLrdbHwhNEv
         mQxkOrmSJpcAeliJ90YDVrPtpaedDjiB/wU1X1fIzNvyzSKiXq+PiMqPLG/RO0ts/p1k
         mk+qTPdsTA4fTP+ByU623uPjzXR5s7+ab66lPqc4YBksEvslMld/VwJ1xVSHiO1YU9pT
         SvHVqPhJHfyB31JUjT53PL8j6/d6BL+FWDYmabOOs+Raz9dnso9QY3U0y68uOhpsJ3q6
         V9/XHUOR+iens0LNunilofFq9W5QKepxiuatbgGoguiRtCJd+50fTlz2eaM50dflIbv8
         sodg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=Qpoq8njNIKPGqHkO3kbGHXyyjepVG6Du3/cu/4cpQMo=;
        b=lNMxje9Z0lT/dob81oKW0ZvMGeugigqGLbpvpy4lc3SWcRmn3NwXdPI3C0+Z2G3X1X
         PFNe0MY/qL8qqf3HuPmi53pWNd+Cy0/P1A1ylyXedLxmM2hTM13/2Bs63vvPhKe/WKdy
         lJOFd1wnfJJdjVg3fexi/m5nm+QdO9VBIW0cu9QJL6HFhg8DedBjX47lwAHAqffMtFmX
         wCVXxFK1uApC6UUDMHA84nTHzCQZFpoAEtHE17B9DmD66e1D7Ns3tEJk52f95fNEYl33
         zZXsxi4mJctfC47DHdCDlrTzxMu9WRz+nw39RClzQMx/DBKrY0mjr/ko3+3+UtcmRcRf
         +9jw==
X-Gm-Message-State: AOUpUlFwQZY2ccep/y3DNSanSugRnxg5XI3xYMY0Ucgw7ULi6mDWHr6v
        bJH18i7bVJqdt1+hR1rL312Q4esnafh1x7dIlMOKASiAgFs7gmD2bY6IrTHCyH+kL1MAA1Vp0M6
        upTmrltVXmfqaaHozPjfYHVKde1okIZ75xpPYrOZ1tv5FXBRN4ApE4i8=
X-Google-Smtp-Source: 
 AA+uWPzqfr6SCWl45H2toynB2Wx4Odg+QKhLMfGHWAfdoaxF7I2RtfR9JRNhHsXpl/vRCLox8rnqj06hXQ==
X-Received: by 2002:aca:f145:: with SMTP id
 p66-v6mr5079041oih.0.1533936373289;
 Fri, 10 Aug 2018 14:26:13 -0700 (PDT)
Date: Fri, 10 Aug 2018 14:26:03 -0700
In-Reply-To: <20180810212604.148650-1-dasch@google.com>
Message-Id: <20180810212604.148650-2-dasch@google.com>
Mime-Version: 1.0
References: <20180810212604.148650-1-dasch@google.com>
X-Mailer: git-send-email 2.18.0.597.ga71716f1ad-goog
Subject: [PATCH 2/3] KVM: x86: Control guest reads of MSR_PLATFORM_INFO
From: Drew Schmitt <dasch@google.com>
To: kvm@vger.kernel.org
Cc: Drew Schmitt <dasch@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Add KVM_CAP_MSR_PLATFORM_INFO so that userspace can disable guest access
to reads of MSR_PLATFORM_INFO.

Signed-off-by: Drew Schmitt <dasch@google.com>
Reviewed-by: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
---
 arch/x86/include/asm/kvm_host.h |  2 ++
 arch/x86/kvm/x86.c              | 10 ++++++++++
 include/uapi/linux/kvm.h        |  1 +
 3 files changed, 13 insertions(+)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index c13cd28d9d1be..c0497c7a4869f 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -853,6 +853,8 @@ struct kvm_arch {
 
 	bool x2apic_format;
 	bool x2apic_broadcast_quirk_disabled;
+
+	bool guest_can_read_msr_platform_info;
 };
 
 struct kvm_vm_stat {
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 0278cde1893ca..b27d88fc6eb8f 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -2743,6 +2743,9 @@ int kvm_get_msr_common(struct kvm_vcpu *vcpu, struct msr_data *msr_info)
 		msr_info->data = vcpu->arch.osvw.status;
 		break;
 	case MSR_PLATFORM_INFO:
+		if (!msr_info->host_initiated &&
+		    !vcpu->kvm->arch.guest_can_read_msr_platform_info)
+			return 1;
 		msr_info->data = vcpu->arch.msr_platform_info;
 		break;
 	case MSR_MISC_FEATURES_ENABLES:
@@ -2890,6 +2893,7 @@ int kvm_vm_ioctl_check_extension(struct kvm *kvm, long ext)
  	case KVM_CAP_SPLIT_IRQCHIP:
 	case KVM_CAP_IMMEDIATE_EXIT:
 	case KVM_CAP_GET_MSR_FEATURES:
+	case KVM_CAP_MSR_PLATFORM_INFO:
 		r = 1;
 		break;
 	case KVM_CAP_SYNC_REGS:
@@ -4259,6 +4263,10 @@ static int kvm_vm_ioctl_enable_cap(struct kvm *kvm,
 			kvm->arch.pause_in_guest = true;
 		r = 0;
 		break;
+	case KVM_CAP_MSR_PLATFORM_INFO:
+		kvm->arch.guest_can_read_msr_platform_info = cap->args[0];
+		r = 0;
+		break;
 	default:
 		r = -EINVAL;
 		break;
@@ -8724,6 +8732,8 @@ int kvm_arch_init_vm(struct kvm *kvm, unsigned long type)
 	kvm->arch.kvmclock_offset = -ktime_get_boot_ns();
 	pvclock_update_vm_gtod_copy(kvm);
 
+	kvm->arch.guest_can_read_msr_platform_info = true;
+
 	INIT_DELAYED_WORK(&kvm->arch.kvmclock_update_work, kvmclock_update_fn);
 	INIT_DELAYED_WORK(&kvm->arch.kvmclock_sync_work, kvmclock_sync_fn);
 
diff --git a/include/uapi/linux/kvm.h b/include/uapi/linux/kvm.h
index b6270a3b38e9f..f3367041b609b 100644
--- a/include/uapi/linux/kvm.h
+++ b/include/uapi/linux/kvm.h
@@ -949,6 +949,7 @@ struct kvm_ppc_resize_hpt {
 #define KVM_CAP_GET_MSR_FEATURES 153
 #define KVM_CAP_HYPERV_EVENTFD 154
 #define KVM_CAP_HYPERV_TLBFLUSH 155
+#define KVM_CAP_MSR_PLATFORM_INFO 156
 
 #ifdef KVM_CAP_IRQ_ROUTING
 

From patchwork Fri Aug 10 21:26:04 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Drew Schmitt <dasch@google.com>
X-Patchwork-Id: 10563151
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id BE81414E2
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:21 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id AF55F2B5B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:21 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id A37BA2B5BF; Fri, 10 Aug 2018 21:26:21 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6E9D92B5B3
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 10 Aug 2018 21:26:20 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727314AbeHJX5y (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 10 Aug 2018 19:57:54 -0400
Received: from mail-ua1-f73.google.com ([209.85.222.73]:56919 "EHLO
        mail-ua1-f73.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727273AbeHJX5y (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 10 Aug 2018 19:57:54 -0400
Received: by mail-ua1-f73.google.com with SMTP id u26-v6so2178945uan.23
        for <kvm@vger.kernel.org>; Fri, 10 Aug 2018 14:26:17 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=GOUflwlDaCR6e3gJ8PsWurMPIvWkOJdhGQA2Efjb2m0=;
        b=eUDZG3qGF+j4lrV3tLz5AIHaFQ96eFO6Uql8tsYSCv3d2v6KQF8Nr+CoGyco5vNdkl
         EZglDfSO3AApCGh7cmhS7KrCeBkg+YU90nBLw7xaNaLeJl1R7IzeKcDnIVcBtxI3GoLq
         sLZaahCcD6d2VjdeGXwNY3zAKnqJH90+iRdjllg2WHgd8Vx9Peq3meoH/KXOnBoSy5JX
         5/yYPDHVMXue4NnGY7L4V6n7Zwg8S8PuQPpjUOnqqStiNqPWRnTx9tszuzAPZEn/6RXG
         JaThCwCeFEK9zZbQ+3w/cYPK1teYqajv3KPPoiMYfE8u2kqywmoLWFIRU2prjXRzi4Ur
         SOzA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=GOUflwlDaCR6e3gJ8PsWurMPIvWkOJdhGQA2Efjb2m0=;
        b=UoUpasZ96J8O7oEVPUNrbvaT8zPh/o3+OeKiyG1wwdeIGtKj+Et8nHei/+Jv6FcyXH
         gccKyDwahEYKxRxOky3xnS0nMW+yHffU+Egl4igOih6Enz4knGofiYTaSCxKet+DVWTg
         IuyAKF9AEZSDrUsZtf1Fb8OgFaEf87k+A9M+tOvrJUnaWdlRcII2iG8FKWNwJVXM85r/
         rWz2nosIIXGaYbvwFVODDUR+EMyLqnmvBLoaV8krJlex2uRFpEEikE0KIbOeLmOb/J73
         h9lqjyVQapDPoTxJwTbPglpInPXpSE3E2iZpj7CpBVqz7ZM1dCjZp5TLzXBYf5QCo5zk
         RJvA==
X-Gm-Message-State: AOUpUlH/oZWY8QhtaaiJhHHR3IbWSUzwTN8X66Ggo2/Wvh0VW9GOSN8d
        4qDcT4GmZObeUWWSMtSGcS7LwGBcJIxZpcLUbKUoZc8gKTeIQM5UUUTl00rwgR61UBZQo9sEwFH
        tNaDSDHKlPjauAu3F6ze8/EdgVoW2QrlHljaboi3UMVtSx0BsMowYhwA=
X-Google-Smtp-Source: 
 AA+uWPwuumqkEeNw2RbKdlcwZWGesN6bKYq+OPk78bC/+slpMraECenAHDFac3ZFULSb4hrXpd+Zgtrc1g==
X-Received: by 2002:a1f:a557:: with SMTP id
 o84-v6mr4449876vke.85.1533936376613;
 Fri, 10 Aug 2018 14:26:16 -0700 (PDT)
Date: Fri, 10 Aug 2018 14:26:04 -0700
In-Reply-To: <20180810212604.148650-1-dasch@google.com>
Message-Id: <20180810212604.148650-3-dasch@google.com>
Mime-Version: 1.0
References: <20180810212604.148650-1-dasch@google.com>
X-Mailer: git-send-email 2.18.0.597.ga71716f1ad-goog
Subject: [PATCH 3/3] kvm: selftests: Add platform_info_test
From: Drew Schmitt <dasch@google.com>
To: kvm@vger.kernel.org
Cc: Drew Schmitt <dasch@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Test guest access to MSR_PLATFORM_INFO when the capability is enabled
or disabled.

Signed-off-by: Drew Schmitt <dasch@google.com>
---
 tools/testing/selftests/kvm/.gitignore        |   1 +
 tools/testing/selftests/kvm/Makefile          |   3 +-
 .../testing/selftests/kvm/include/kvm_util.h  |   4 +
 tools/testing/selftests/kvm/lib/kvm_util.c    |  89 +++++++++++++
 .../selftests/kvm/platform_info_test.c        | 121 ++++++++++++++++++
 5 files changed, 217 insertions(+), 1 deletion(-)
 create mode 100644 tools/testing/selftests/kvm/platform_info_test.c

diff --git a/tools/testing/selftests/kvm/.gitignore b/tools/testing/selftests/kvm/.gitignore
index 63fc1ab9248fa..8078891ded84a 100644
--- a/tools/testing/selftests/kvm/.gitignore
+++ b/tools/testing/selftests/kvm/.gitignore
@@ -1,3 +1,4 @@
+platform_info_test
 set_sregs_test
 sync_regs_test
 vmx_tsc_adjust_test
diff --git a/tools/testing/selftests/kvm/Makefile b/tools/testing/selftests/kvm/Makefile
index d9d00319b07cd..85e3eb4108c28 100644
--- a/tools/testing/selftests/kvm/Makefile
+++ b/tools/testing/selftests/kvm/Makefile
@@ -6,7 +6,8 @@ UNAME_M := $(shell uname -m)
 LIBKVM = lib/assert.c lib/elf.c lib/io.c lib/kvm_util.c lib/sparsebit.c
 LIBKVM_x86_64 = lib/x86.c lib/vmx.c
 
-TEST_GEN_PROGS_x86_64 = set_sregs_test
+TEST_GEN_PROGS_x86_64 = platform_info_test
+TEST_GEN_PROGS_x86_64 += set_sregs_test
 TEST_GEN_PROGS_x86_64 += sync_regs_test
 TEST_GEN_PROGS_x86_64 += vmx_tsc_adjust_test
 
diff --git a/tools/testing/selftests/kvm/include/kvm_util.h b/tools/testing/selftests/kvm/include/kvm_util.h
index 637b7017b6ee5..a7e34eaccadba 100644
--- a/tools/testing/selftests/kvm/include/kvm_util.h
+++ b/tools/testing/selftests/kvm/include/kvm_util.h
@@ -50,6 +50,7 @@ enum vm_mem_backing_src_type {
 };
 
 int kvm_check_cap(long cap);
+int vm_enable_cap(struct kvm_vm *vm, struct kvm_enable_cap *cap);
 
 struct kvm_vm *vm_create(enum vm_guest_mode mode, uint64_t phy_pages, int perm);
 void kvm_vm_free(struct kvm_vm *vmp);
@@ -103,6 +104,9 @@ void vcpu_events_get(struct kvm_vm *vm, uint32_t vcpuid,
 			  struct kvm_vcpu_events *events);
 void vcpu_events_set(struct kvm_vm *vm, uint32_t vcpuid,
 			  struct kvm_vcpu_events *events);
+uint64_t vcpu_get_msr(struct kvm_vm *vm, uint32_t vcpuid, uint64_t msr_index);
+void vcpu_set_msr(struct kvm_vm *vm, uint32_t vcpuid, uint64_t msr_index,
+	uint64_t msr_value);
 
 const char *exit_reason_str(unsigned int exit_reason);
 
diff --git a/tools/testing/selftests/kvm/lib/kvm_util.c b/tools/testing/selftests/kvm/lib/kvm_util.c
index 37e2a787d2fcc..2d13b26d32215 100644
--- a/tools/testing/selftests/kvm/lib/kvm_util.c
+++ b/tools/testing/selftests/kvm/lib/kvm_util.c
@@ -62,6 +62,29 @@ int kvm_check_cap(long cap)
 	return ret;
 }
 
+/* VM Enable Capability
+ *
+ * Input Args:
+ *   vm - Virtual Machine
+ *   cap - Capability
+ *
+ * Output Args: None
+ *
+ * Return: On success, 0. On failure a TEST_ASSERT failure is produced.
+ *
+ * Enables a capability (KVM_CAP_*) on the VM.
+ */
+int vm_enable_cap(struct kvm_vm *vm, struct kvm_enable_cap *cap)
+{
+	int ret;
+
+	ret = ioctl(vm->fd, KVM_ENABLE_CAP, cap);
+	TEST_ASSERT(ret == 0, "KVM_ENABLE_CAP IOCTL failed,\n"
+		"  rc: %i errno: %i", ret, errno);
+
+	return ret;
+}
+
 /* VM Create
  *
  * Input Args:
@@ -1124,6 +1147,72 @@ void vcpu_events_set(struct kvm_vm *vm, uint32_t vcpuid,
 		ret, errno);
 }
 
+/* VCPU Get MSR
+ *
+ * Input Args:
+ *   vm - Virtual Machine
+ *   vcpuid - VCPU ID
+ *   msr_index - Index of MSR
+ *
+ * Output Args: None
+ *
+ * Return: On success, value of the MSR. On failure a TEST_ASSERT is produced.
+ *
+ * Get value of MSR for VCPU.
+ */
+uint64_t vcpu_get_msr(struct kvm_vm *vm, uint32_t vcpuid, uint64_t msr_index)
+{
+	struct vcpu *vcpu = vcpu_find(vm, vcpuid);
+	struct {
+		struct kvm_msrs header;
+		struct kvm_msr_entry entry;
+	} buffer = {};
+	int r;
+
+	TEST_ASSERT(vcpu != NULL, "vcpu not found, vcpuid: %u", vcpuid);
+	buffer.header.nmsrs = 1;
+	buffer.entry.index = msr_index;
+	r = ioctl(vcpu->fd, KVM_GET_MSRS, &buffer.header);
+	TEST_ASSERT(r == 1, "KVM_GET_MSRS IOCTL failed,\n"
+		"  rc: %i errno: %i", r, errno);
+
+	return buffer.entry.data;
+}
+
+/* VCPU Set MSR
+ *
+ * Input Args:
+ *   vm - Virtual Machine
+ *   vcpuid - VCPU ID
+ *   msr_index - Index of MSR
+ *   msr_value - New value of MSR
+ *
+ * Output Args: None
+ *
+ * Return: On success, nothing. On failure a TEST_ASSERT is produced.
+ *
+ * Set value of MSR for VCPU.
+ */
+void vcpu_set_msr(struct kvm_vm *vm, uint32_t vcpuid, uint64_t msr_index,
+	uint64_t msr_value)
+{
+	struct vcpu *vcpu = vcpu_find(vm, vcpuid);
+	struct {
+		struct kvm_msrs header;
+		struct kvm_msr_entry entry;
+	} buffer = {};
+	int r;
+
+	TEST_ASSERT(vcpu != NULL, "vcpu not found, vcpuid: %u", vcpuid);
+	memset(&buffer, 0, sizeof(buffer));
+	buffer.header.nmsrs = 1;
+	buffer.entry.index = msr_index;
+	buffer.entry.data = msr_value;
+	r = ioctl(vcpu->fd, KVM_SET_MSRS, &buffer.header);
+	TEST_ASSERT(r == 1, "KVM_SET_MSRS IOCTL failed,\n"
+		"  rc: %i errno: %i", r, errno);
+}
+
 /* VM VCPU Args Set
  *
  * Input Args:
diff --git a/tools/testing/selftests/kvm/platform_info_test.c b/tools/testing/selftests/kvm/platform_info_test.c
new file mode 100644
index 0000000000000..3c14959a09210
--- /dev/null
+++ b/tools/testing/selftests/kvm/platform_info_test.c
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Test for x86 KVM_CAP_MSR_PLATFORM_INFO
+ *
+ * Copyright (C) 2018, Google LLC.
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2.
+ *
+ * Verifies expected behavior of controlling guest access to
+ * MSR_PLATFORM_INFO.
+ */
+
+#define _GNU_SOURCE /* for program_invocation_short_name */
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+
+#include "test_util.h"
+#include "kvm_util.h"
+#include "x86.h"
+
+#define VCPU_ID 0
+#define PORT_HOST_SYNC 0x1000
+#define MSR_PLATFORM_INFO_MAX_TURBO_RATIO 0xff00
+
+static void __exit_to_l0(uint16_t port, unsigned long arg)
+{
+	__asm__ __volatile__("in %[port], %%al"
+		:
+		: [port]"d"(port), "D"(arg)
+		: "rax");
+}
+
+#define exit_to_l0(_port, _arg) __exit_to_l0(_port, (unsigned long) (_arg))
+
+static void guest_code(void)
+{
+	uint64_t msr_platform_info;
+
+	for (;;) {
+		msr_platform_info = rdmsr(MSR_PLATFORM_INFO);
+		exit_to_l0(PORT_HOST_SYNC, msr_platform_info);
+		asm volatile ("inc %r11");
+	}
+}
+
+static void set_msr_platform_info_enabled(struct kvm_vm *vm, bool enable)
+{
+	struct kvm_enable_cap cap = {};
+
+	cap.cap = KVM_CAP_MSR_PLATFORM_INFO;
+	cap.flags = 0;
+	cap.args[0] = (int)enable;
+	vm_enable_cap(vm, &cap);
+}
+
+static void test_msr_platform_info_enabled(struct kvm_vm *vm)
+{
+	struct kvm_run *run = vcpu_state(vm, VCPU_ID);
+	struct kvm_regs regs;
+
+	set_msr_platform_info_enabled(vm, true);
+	vcpu_run(vm, VCPU_ID);
+	TEST_ASSERT(run->exit_reason == KVM_EXIT_IO,
+			"Exit_reason other than KVM_EXIT_IO: %u (%s),\n",
+			run->exit_reason,
+			exit_reason_str(run->exit_reason));
+	TEST_ASSERT(run->io.port == PORT_HOST_SYNC,
+			"Received IO from port other than PORT_HOST_SYNC: %u\n",
+			run->io.port);
+	vcpu_regs_get(vm, VCPU_ID, &regs);
+	TEST_ASSERT((regs.rdi & MSR_PLATFORM_INFO_MAX_TURBO_RATIO) ==
+		MSR_PLATFORM_INFO_MAX_TURBO_RATIO,
+		"Expected MSR_PLATFORM_INFO to have max turbo ratio mask: %i.",
+		MSR_PLATFORM_INFO_MAX_TURBO_RATIO);
+}
+
+static void test_msr_platform_info_disabled(struct kvm_vm *vm)
+{
+	struct kvm_run *run = vcpu_state(vm, VCPU_ID);
+
+	set_msr_platform_info_enabled(vm, false);
+	vcpu_run(vm, VCPU_ID);
+	TEST_ASSERT(run->exit_reason == KVM_EXIT_SHUTDOWN,
+			"Exit_reason other than KVM_EXIT_SHUTDOWN: %u (%s)\n",
+			run->exit_reason,
+			exit_reason_str(run->exit_reason));
+}
+
+int main(int argc, char *argv[])
+{
+	struct kvm_vm *vm;
+	struct kvm_run *state;
+	int rv;
+	uint64_t msr_platform_info;
+
+	/* Tell stdout not to buffer its content */
+	setbuf(stdout, NULL);
+
+	rv = kvm_check_cap(KVM_CAP_MSR_PLATFORM_INFO);
+	if (!rv) {
+		fprintf(stderr,
+			"KVM_CAP_MSR_PLATFORM_INFO not supported, skip test\n");
+		exit(KSFT_SKIP);
+	}
+
+	vm = vm_create_default(VCPU_ID, guest_code);
+
+	msr_platform_info = vcpu_get_msr(vm, VCPU_ID, MSR_PLATFORM_INFO);
+	vcpu_set_msr(vm, VCPU_ID, MSR_PLATFORM_INFO,
+		msr_platform_info | MSR_PLATFORM_INFO_MAX_TURBO_RATIO);
+	test_msr_platform_info_disabled(vm);
+	test_msr_platform_info_enabled(vm);
+	vcpu_set_msr(vm, VCPU_ID, MSR_PLATFORM_INFO, msr_platform_info);
+
+	kvm_vm_free(vm);
+
+	return 0;
+}
